types/react-dom/index.d.ts

```diff
@@ -12,13 +12,25 @@ export as namespace ReactDOM;
 import {
     ReactInstance, Component, ComponentState,
     ReactElement, SFCElement, CElement,
-    DOMAttributes, DOMElement
+    DOMAttributes, DOMElement, ReactNode
 } from 'react';
 
 export function findDOMNode<E extends Element>(instance: ReactInstance): E;
 export function findDOMNode(instance: ReactInstance): Element;
 export function unmountComponentAtNode(container: Element): boolean;
 
+// @TODO: I think ReactPortal should be part of React types (as a valid ReactNode)
+// Mostly copied from Reacts own flow definitions:
+// https://github.com/facebook/react/blob/589c0a25dfa18c2090549cc6f5b626d69ea53c2a/src/renderers/shared/fiber/isomorphic/ReactTypes.js#L43-L50
+export interface ReactPortal {
+    $$typeof: symbol | number;
+    key: string | null;
+    containerInfo: any;
+    children: ReactNode;
+    implementation: any;
+}
+export function createPortal(children: ReactNode, container: Element, key?: string): ReactPortal;
```

Comment:
The return type here is tricky because the way we create portals is by returning them from a `render` function, but component classes/functions are expected to return `JSX.Element`.

E.g.,

```ts
interface Props {
    isOpen: boolean;
}

class Modal extends React.Component<Props> {
    render(): JSX.Element | null {
        // compile error: React.ReactPortal is not assignable to JSX.Element
        return this.props.isOpen ? ReactDOM.createPortal(...) : null;
    }
}
```

There are 2 solutions I can think of:

1. (less invasive) - make the return type of `createPortal` an intersection: `React.ReactNode & JSX.Element`
2. (more invasive) - change the return type for `render` and `StatelessComponent` to be `JSX.Element | ReactPortal | null`

I'm not sure which is better, but maybe it's worth adding tests to make sure `createPortal` is compatible with render functions.

Reply:
Actually, the more I think about it, the return type for `render` should be widened (option 2 above). It's even in the [React docs](https://reactjs.org/docs/react-component.html#render):

> When called, it should examine `this.props` and `this.state` and return one of the following types:
> - React elements...
> - String and numbers. These are rendered as text nodes in the DOM.
> - Portals. Created with ReactDOM.createPortal.
> - `null`. Renders nothing.
> - Booleans. Render nothing...

Per the above, the return type should be:

```ts
render(): JSX.Element | JSX.Element[] | ReactPortal | string | number | null | false;
```

This should also be the return type for `StatelessComponent`'s invocation signature.

Reply:
Thanks for the guidance, is it appropriate to make the changes to `types/react` in this PR?

Reply:
@theruther4d Yes, I think that would certainly be appropriate since the changes are required/related.

Reply:
@jgoz Updated per your review. For class-based components everything works as expected, but I can't seem to figure out how to allow `ReactPortal` as the return type for `StatelessComponent`. Updating the invocation signature to `JSX.Element | JSX.Element[] | ReactPortal | string | number | null | false;` causes issues. Can you point me in the right direction?

Reply:
@theruther4d what exactly are the issues that you hit?